liquidator/OneInchExchange.js

```diff
@@ -59,38 +90,40 @@ class OneInchExchange {
       });
     }
 
-    // 1 Split config
-    const flags = 0; // Enables all exchanges
-
-    // Number of pieces source volume could be splitted
-    // (Works like granularity, higly affects gas usage.
-    // Should be called offchain, but could be called onchain
-    // if user swaps not his own funds, but this is still considered
-    // as not safe)
-    // More info: https://github.com/CryptoManiacsZone/1inchProtocol#getexpectedreturn
-    const parts = 2;
-
     const expectedReturn = await this.oneSplitContract.methods
-      .getExpectedReturn(fromToken, toToken, amountWei, parts, flags)
+      .getExpectedReturn(fromToken, toToken, amountWei, this.oneInchParts, this.oneInchFlags)
       .call();
 
     const { returnAmount, distribution } = expectedReturn;
 
     this.logger.debug({
       at: "OneInchExchange",
       message: "GetExpectedReturn",
+      currentTime,
       returnAmount,
       distribution
     });
 
+    if (minReturnAmountWei && toBN(returnAmount.toString()).lt(toBN(minReturnAmountWei.toString()))) {
+      this.logger.debug({
+        at: "OneInchExchange",
+        message: "ReturnAmountTooLow",
+        currentTime,
+        returnAmount,
+        minReturnAmountWei
+      });
+      throw new Error("OneInch return amount too low");
```

Comment:
I don't think we want to throw an `Error` in combination with a log like this. If something error out then simply throw the error. or, use the error level. See sending [here](https://github.com/UMAprotocol/protocol/blob/dccb2feb9e5aa883648ee7dab4896c1bc6d92694/financial-templates-lib/price-feed/CreatePriceFeed.js#L102) for an example error log with Winston.

In either case, is this the kind of behavior we want wit this kind of error? I would think it better to make this an `logger.info` type event or `logger.warn`. We want to know about it but it should surely not crash the bot? Throwing an error like this would bubble all the way up to the main try-catch and cause the bot to re-try the whole execution loop. Before we had re-try logic it would yield the process. In either case this is too aggressive for this kind of error and it should be dealt with more gracefully I think.

Reply:
With the retry logic this might actually be useful to throw an error here, wdyt @chrismaree ? But what doesn't sit right with me is why throw an error here and not other places?

Reply:
@chrismaree can you clarify on this?

Reply:
@kendricktan if the promise is rejected but something went fatally wrong then it should _still_ not send a debug message. A debug message is not readable by anything other than directly from the GCP console. If we want to see it in slack or pagerduty it needs to be _at least_ an `info` level message. Throwing after a debug means that the bot would fail silently on the actual issue and then be caught upstream by the main try-catch. This is not expected behavior and breaks how we deal with errors in other places. If you want to throw an error then the best practice is to remove the winston log, throw the error and then let it be caught and logged upstream. You can see a detailed breakdown of the logging levels [here](https://github.com/UMAprotocol/protocol/blob/8aae10fee7c944351f792cf72c924899ad9fa216/packages/financial-templates-lib/src/logger/Logger.js#L1).

To see how we deal with other fatal errors (like the bot not having a valid price feed) check out [here](https://github.com/UMAprotocol/protocol/blob/8aae10fee7c944351f792cf72c924899ad9fa216/packages/liquidator/index.js#L73). This is effectively the same situation: an error occurred that is so bad that the process MUST yield. it can't continue.

My concern though is that this does not seem like a fatal error. The bot requested a quote from one inch and the amount returned was less than the `minReturnAmountWei`. Does this mean the bot should fall over? I dont think so.

My recommendation here: 1) change the log level to `warn`. this will inform us that there is a problem with the 1inch estimation. 2) replace this throw error with a `return` so the process yields before trying to send the transaction.

What do you think?